graph/db/sql_store.go

```diff
@@ -1549,3 +1669,370 @@ func upsertChanPolicyExtraSignedFields(ctx context.Context, db SQLQueries,
+	chain chainhash.Hash, dbChanID int64, dbChanRow any, node1,
+	node2 route.Vertex) (*models.ChannelEdgeInfo, error) {
+
+	dbChan, err := extractChannel(dbChanRow)
+	if err != nil {
+		return nil, err
+	}
+
+	fv, extras, err := getChanFeaturesAndExtras(
+		ctx, db, dbChanID,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	op, err := wire.NewOutPointFromString(dbChan.Outpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	var featureBuf bytes.Buffer
+	if err := fv.Encode(&featureBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode features: %w", err)
+	}
+
+	recs, err := lnwire.CustomRecords(extras).Serialize()
+	if err != nil {
+		return nil, fmt.Errorf("unable to serialize extra signed "+
+			"fields: %w", err)
+	}
+	if recs == nil {
+		recs = make([]byte, 0)
+	}
+
+	var btcKey1, btcKey2 route.Vertex
+	copy(btcKey1[:], dbChan.BitcoinKey1)
+	copy(btcKey2[:], dbChan.BitcoinKey2)
+
+	channel := &models.ChannelEdgeInfo{
+		ChainHash:        chain,
+		ChannelID:        byteOrder.Uint64(dbChan.Scid),
+		NodeKey1Bytes:    node1,
+		NodeKey2Bytes:    node2,
+		BitcoinKey1Bytes: btcKey1,
+		BitcoinKey2Bytes: btcKey2,
+		ChannelPoint:     *op,
+		Capacity:         btcutil.Amount(dbChan.Capacity.Int64),
+		Features:         featureBuf.Bytes(),
+		ExtraOpaqueData:  recs,
+	}
+
+	if dbChan.Bitcoin1Signature != nil {
+		channel.AuthProof = &models.ChannelAuthProof{
+			NodeSig1Bytes:    dbChan.Node1Signature,
+			NodeSig2Bytes:    dbChan.Node2Signature,
+			BitcoinSig1Bytes: dbChan.Bitcoin1Signature,
+			BitcoinSig2Bytes: dbChan.Bitcoin2Signature,
+		}
+	}
+
+	return channel, nil
+}
+
+// buildNodeVertices is a helper that converts raw node public keys
+// into route.Vertex instances.
+func buildNodeVertices(node1Pub, node2Pub []byte) (route.Vertex,
+	route.Vertex, error) {
+
+	node1Vertex, err := route.NewVertexFromBytes(node1Pub)
+	if err != nil {
+		return route.Vertex{}, route.Vertex{}, fmt.Errorf("unable to "+
+			"create vertex from node1 pubkey: %w", err)
+	}
+
+	node2Vertex, err := route.NewVertexFromBytes(node2Pub)
+	if err != nil {
+		return route.Vertex{}, route.Vertex{}, fmt.Errorf("unable to "+
+			"create vertex from node2 pubkey: %w", err)
+	}
+
+	return node1Vertex, node2Vertex, nil
+}
+
+// getChanFeaturesAndExtras fetches the channel features and extra TLV types
+// for a channel with the given ID.
+func getChanFeaturesAndExtras(ctx context.Context, db SQLQueries,
+	id int64) (*lnwire.FeatureVector, map[uint64][]byte, error) {
+
+	rows, err := db.GetChannelFeaturesAndExtras(ctx, id)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to fetch channel "+
+			"features and extras: %w", err)
+	}
+
+	var (
+		fv     = lnwire.EmptyFeatureVector()
+		extras = make(map[uint64][]byte)
+	)
+	for _, row := range rows {
+		switch row.Kind {
+		case "feature":
+			featureBit, err := strconv.Atoi(row.Key)
+			if err != nil {
+				return nil, nil, err
+			}
+			fv.Set(lnwire.FeatureBit(featureBit))
+
+		case "extra":
```

Comment:
nit: to make this function more robust against future changes to the underlying query, i'd recommend adding a default case as sanity check.

Reply:
good catch! added